source/sessions/driver-sessions.rst

```diff
@@ -620,10 +620,11 @@ A session ID MUST NOT be used simultaneously by more than one operation. Since
 drivers don't wait for a response for an unacknowledged write a driver would
 not know when the session ID could be reused. In theory a driver could use a
 new session ID for each unacknowledged write, but that would result in many
-orphaned sessions building up at the server. Rather than prohibit using
-unacknowledged writes with sessions we choose instead to specify that a driver
-MUST NOT send a session ID with unacknowledged writes. This is true for both
-implicit and explicit sessions.
+orphaned sessions building up at the server. A driver MUST NOT send a session
+ID with unacknowledged writes. Drivers SHOULD raise an error if users request
+an unacknowledged write with a session.  Drivers MUST document the behavior
+of unacknowledged writes with a session.  This is true for both implicit and
+explicit sessions.
```

Comment:
If we are going to prohibit unacknowledged writes with a session we may have to make an exception for implicit sessions. Otherwise we are basically just prohibiting unacknowledged writes.

Reply:
I think the idea discussed by the leads in the spec triage meeting was that since sessions mandate client-serialized operations and unack'd writes can't provide that guarantee, any unack'ed writes must be --by definition-- outside a session.  I.e. never part of an implicit session and an error if requested explicitly.

Reply:
But there's no such thing as an operation that is not part of an implicit session. The lack of an explicit session is by definition a use of an implicit session. Our APIs have no way for the user to say do NOT use an implicit session.

Reply:
> Our APIs have no way for the user to say do NOT use an implicit session.

Is there reason we cannot say that an unacknowledged write concern implies that the operation will not include an `lsid` (i.e. no implicit session)? Furthermore, it's a client-side logical error if the user supplies an explicit session for an unacknowledged write.

This seems comparable to the retryable writes spec saying that unacknowledged writes will not get a transaction ID added to their command document.

Reply:
Proposed change:

A driver MUST NOT send a session ID with unacknowledged writes. Drivers MUST raise an error if users request an unacknowledged write with an explicit session. Unacknowledged writes are only permitted with implicit sessions, in which case the command sent to the server MUST NOT include the session ID.

Reply:
Either that, or the spec has to be changed further. Currently the spec says that every operation  that is not called with an explicit session automatically starts an implicit session for that one operation. If we don't allow unacknowledged writes with implicit sessions then we need to have more rules about whether an implicit session is started or not.

Reply:
You said:
>A driver MUST NOT send a session ID with unacknowledged writes. Drivers MUST raise an error if users request an unacknowledged write with an explicit session. Unacknowledged writes are only permitted with implicit sessions, in which case the command sent to the server MUST NOT include the session ID.

To me, that's essentially what I said except I said "SHOULD" because drivers can't introduce a new error and break code -- which is why I said they "MUST" document what they do.

I'll try another iteration and you can tell me if it reads better to you.